agent/src/main/java/com/microsoft/hydralab/agent/runner/ActionExecutor.java

```diff
@@ -0,0 +1,98 @@
+// Copyright (c) Microsoft Corporation.
+// Licensed under the MIT License.
+package com.microsoft.hydralab.agent.runner;
+
+import com.alibaba.fastjson.JSONObject;
+import com.microsoft.hydralab.common.entity.common.DeviceAction;
+import com.microsoft.hydralab.common.entity.common.DeviceInfo;
+import com.microsoft.hydralab.common.management.DeviceManager;
+import com.microsoft.hydralab.common.util.ThreadUtils;
+import org.jetbrains.annotations.NotNull;
+import org.slf4j.Logger;
+
+import java.lang.reflect.InvocationTargetException;
+import java.lang.reflect.Method;
+import java.util.Arrays;
+import java.util.Comparator;
+import java.util.List;
+import java.util.Set;
+import java.util.stream.Collectors;
+
+/**
+ * @author zhoule
+ * @date 12/20/2022
+ */
+
+public class ActionExecutor {
+    /**
+     * the implementation of supported actions should not be overload
+     */
+    private Set<String> actionTypes = Set.of("setProperty", "setDefaultLauncher", "backToHome");
+
+    public void doActions(@NotNull DeviceManager deviceManager, @NotNull DeviceInfo deviceInfo, @NotNull Logger logger,
+                          @NotNull List<DeviceAction> actions, @NotNull String timing) {
+        if (actions.isEmpty()) {
+            return;
+        }
+        //filter todoActions
+        //sort: asc
+        List<DeviceAction> todoActions = actions.stream().filter(deviceAction -> timing.equals(deviceAction.getTimingToAct()))
+                .filter(deviceAction -> actionTypes.contains(deviceAction.getActionType()))
+                .sorted(Comparator.comparingInt(DeviceAction::getActionOrder)).collect(Collectors.toList());
+
+        logger.info("Start to execute actions! Current timing is {}, action size is {}", timing, todoActions.size());
+        todoActions.forEach(deviceAction -> doAction(deviceManager, deviceInfo, logger, deviceAction));
+        logger.info("Execute actions finished!");
+        ThreadUtils.safeSleep(3000);
+    }
+
+    public void doAction(@NotNull DeviceManager deviceManager, @NotNull DeviceInfo deviceInfo, @NotNull Logger logger,
+                         @NotNull DeviceAction deviceAction) {
+        try {
+            if (!actionTypes.contains(deviceAction.getActionType())) {
+                return;
+            }
+            logger.info("Start to analysis action type! Current action is {}, order is {}", deviceAction.getActionType(), deviceAction.getActionOrder());
```

Comment:
Why do we need action order? I think it's already an array with sequence, right?